File: src/main/java/org/apache/prepbuddy/cleansers/imputation/NaiveBayesSubstitution.java
package org.apache.prepbuddy.cleansers.imputation;

import org.apache.prepbuddy.rdds.TransformableRDD;
import org.apache.prepbuddy.transformers.RowPurger;
import org.apache.prepbuddy.utils.RowRecord;

/**
 * An imputation strategy that is based on Naive Bayes Algorithm which
 * is the probabilistic classifier. This implementation is only
 * for imputing the categorical values.
 */
public class NaiveBayesSubstitution implements ImputationStrategy {

    private NaiveBayesClassifier naiveBayesClassifier;
    private int[] otherColumns;

    public NaiveBayesSubstitution(int... independentColumnIndexes) {
        otherColumns = independentColumnIndexes;
    }

    @Override
    public void prepareSubstitute(TransformableRDD rdd, final int missingDataColumn) {
        naiveBayesClassifier = new NaiveBayesClassifier(missingDataColumn, otherColumns);
        TransformableRDD trainingSet = rdd.removeRows(new RowPurger.Predicate() {
            @Override
            public Boolean evaluate(RowRecord record) {
                return record.hasEmptyColumn();

            }
        });

        naiveBayesClassifier.train(trainingSet);
    }


    @Override
    public String handleMissingData(RowRecord record) {
        return naiveBayesClassifier.classify(record);
    }

}

File: src/main/scala/org/apache/datacommons/prepbuddy/imputations/NaiveBayesSubstitution.scala
package org.apache.datacommons.prepbuddy.imputations

import org.apache.datacommons.prepbuddy.clusterers.TextFacets
import org.apache.datacommons.prepbuddy.rdds.TransformableRDD
import org.apache.datacommons.prepbuddy.utils.{NumberMap, PivotTable, Probability, RowRecord}

/**
  * An imputation strategy that is based on Naive Bayes Algorithm which
  * is the probabilistic classifier. This implementation is only
  * for imputing the categorical values.
  */
class NaiveBayesSubstitution(independentColumnIndexes: Array[Int]) extends ImputationStrategy {
    private var probs: PivotTable[Probability] = null
    private var permissibleValues: Array[String] = null

    override def prepareSubstitute(rdd: TransformableRDD, missingDataColumn: Int): Unit = {
        val trainingSet: TransformableRDD = rdd.removeRows(_.hasEmptyColumn)
        val facets: TextFacets = trainingSet.listFacets(missingDataColumn)
        permissibleValues = facets.cardinalValues
        val rowKeys: Array[(String, Int)] = facets.rdd.collect()
        val frequencyTable: PivotTable[Integer] = {
            rdd.pivotByCount(missingDataColumn, independentColumnIndexes)
        }
        val totalRows: Long = trainingSet.count()

        probs = frequencyTable.transform((eachValue) => {
            val oneProbability: Double = eachValue.toString.toDouble / totalRows
            new Probability(oneProbability)
        }, new Probability(0)).asInstanceOf[PivotTable[Probability]]

        rowKeys.foreach((each) => {
            val probability: Probability = new Probability(each._2.toString.toDouble / totalRows)
            probs.addEntry(each._1, each._1, probability)
        })

    }

    override def handleMissingData(record: RowRecord): String = {
        val numbers: NumberMap = new NumberMap()
        for (permissibleValue <- permissibleValues) {
            var naiveProbability: Probability = new Probability(1)
            val permissibleProb: Probability = probs.valueAt(permissibleValue, permissibleValue)
            independentColumnIndexes.foreach((columnIndex) => {
                val columnValue: String = record.select(columnIndex).trim
                val probability: Probability = probs.valueAt(permissibleValue, columnValue)
                naiveProbability = naiveProbability.multiply(probability.divide(permissibleProb))
            })
            naiveProbability = naiveProbability.multiply(permissibleProb)
            numbers.put(permissibleValue, naiveProbability.doubleValue)
        }
        numbers.keyWithHighestValue
    }
}

